### ISSUE TITLE
pin_versions_prune removes the newest, not the oldest versions

### ISSUE THREAD

**Issue by tomsing1** [rstudio/pins-r#637]
I think the `pin_versions_prune` function is removing the youngest, not the oldest versions of a pin when the `n` argument is set:

```r
library(pins)

board <- board_temp(versioned = TRUE)

board %>% pin_write(data.frame(x = 1:5), name = "df")
Sys.sleep(1)
board %>% pin_write(data.frame(x = 2:6), name = "df")
Sys.sleep(1)
board %>% pin_write(data.frame(x = 3:7), name = "df")
Sys.sleep(1)

pin_versions_prune(board, "df", 1)
```

**Comment by tomsing1** [comment]
Looking at how the versions to keep are identified in the https://github.com/rstudio/pins-r/blob/9e8c08123861eeef4853caa5ffd1970c9be0457e/R/pin_versions.R#L110 function, it seems that the versions are ordered by the `created` POSIX timestamp. 

At least on my system, that field doesn't have enough precision to distinguish between versions written within a one minute interval (e.g. all of the `created` fields in my example are the same `16:41:00` even though I ensured that they were written at least 1s apart).

But the `version` field has precision down to the second and ordering by this field would return the expected result. Because the `version` is also used to uniquely identify a file, perhaps the `versions_keep` function should receive the `version` as input, not `created`?

**Comment by tomsing1** [comment]
Actually, is there a reason that the time stamp does not have second-level precision?
It seems the `parse_8601_compact` and `parse_8601 ` functions don't contain the `%S` part of the format string. Maybe that should be added to avoid losing that information? 

https://github.com/rstudio/pins-r/blob/9e8c08123861eeef4853caa5ffd1970c9be0457e/R/meta.R#L51

```r
> as.POSIXct(strptime(x, "%Y%m%dT%H%M", tz = "UTC"))
[1] "2022-08-18 00:05:00 UTC"
```
could be
```r
> as.POSIXct(strptime(x, "%Y%m%dT%H%M%S", tz = "UTC"))
[1] "2022-08-18 00:05:58 UTC"
```

That would make ordering versions by the `created` field work at least for versions created within > 1s of each other?

**Comment by tomsing1** [comment]
Interesting - as I was writing more of my unit tests, I noticed that the `pin_write` function behaves differently when it is run within or outside unit tests (of the testthat variety). That's because the time stamp part of the `version`  is fixed in a unit test:

https://github.com/rstudio/pins-r/blob/9e8c08123861eeef4853caa5ffd1970c9be0457e/R/pin_versions.R#L146

I didn't expect that, and it took me a good while to figure out why my unit tests were failing. (I was writing the same file twice into a versioned board. My project includes tracking the metadata in a SQL database - and that complained about duplicated `version` keys. 

There is nothing intrinsically wrong with the current approach, but it reminds me a bit of the [Volkswagen scandal]() and it definitely was surprising. (I solved it by temporarily unsetting the `TESTTHAT` environmental variable in my test.)

**Comment by tomsing1** [comment]
After exploring how pins sets the `version` tag, I am wondering if perhaps the construction of `timestamp+partial hash` is a bit too risky. Maybe generating a `uuid` would be safer? That way, there wouldn't be a need to worry about the precision of the timestamp, either.

**Comment by juliasilge** [comment]
Let's track in this issue the problem folks face when their pins are updated quickly (one minute doesn't seem _that_ quick honestly) and our current implementation of `pin_versions()` cannot find the latest version. 

I'll open a new issue about our use of `is_testing()`. 👍

**Comment by iainmwallace** [comment]
I agree with suggestion of a uuid for versions as opposed to the current time stamp approach. Time of version creation would be an important component to maintain.

A colleague mentioned they ran into issues if they quickly modified/corrected metadata associated with a pin which might have been due to this issue of an update in less than a minute

**Comment by tomsing1** [comment]
I just learned about [ulid](https://cran.r-project.org/web/packages/ulid/vignettes/intro-to-ulid.html) as an alternative to uuids.  `ulid`'s have several advantages - including the fact that they can be lexicographically sorted. That might be help with identifying the most recent / order of pins?

**Comment by machow** [comment]
I wonder if the original issue has been resolved? As I understand it, it was...

* second level info not in the created field of `pin_versions()`.
* `pins_versions_prune()` deletes wrong versions.

Reprex re-running the original code (I see second level info in the timestamps now):

``` r
library(pins)
board <- board_temp(versioned = TRUE)
board %>% pin_write(data.frame(x = 1:5), name = "df")
#> Guessing `type = 'rds'`
#> Creating new version '20230106T185411Z-1fc6c'
#> Writing to pin 'df'
Sys.sleep(1)
board %>% pin_write(data.frame(x = 2:6), name = "df")
#> Guessing `type = 'rds'`
#> Creating new version '20230106T185412Z-a9b87'
#> Writing to pin 'df'
Sys.sleep(1)
board %>% pin_write(data.frame(x = 3:7), name = "df")
#> Guessing `type = 'rds'`
#> Creating new version '20230106T185413Z-dd750'
#> Writing to pin 'df'
Sys.sleep(1)

pin_versions(board, "df")
#> # A tibble: 3 × 3
#>   version                created             hash 
#>   <chr>                  <dttm>              <chr>
#> 1 20230106T185411Z-1fc6c 2023-01-06 13:54:11 1fc6c
#> 2 20230106T185412Z-a9b87 2023-01-06 13:54:12 a9b87
#> 3 20230106T185413Z-dd750 2023-01-06 13:54:13 dd750
```

<sup>Created on 2023-01-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by tomsing1** [comment]
@juliasilge Yes, in my case 1-second precision is sufficient. (I only realized that prior to the fix the precision was 1-minute when I was writing unit tests for a package built on top of `pins`. Real use cases won't write that frequently / quickly.)

I only mentioned the `ulid` package here because it seemed like a principled solution to creating a sortable hash, potentially obviating the need for concatenating the `created` and `hash` fields altogether. (Those have their own uses, of course.)

**Comment by juliasilge** [comment]
Yes, if we end up pursuing a different strategy altogether for identifying versions (which would solve #589) then something like ULIDs could be a good option.

For now, let's say that the original issue with identifying the latest version is no longer a problem. 👍
